### PR TITLE
[D1] Backport __ctfe

### PR DESCRIPTION
**Pull request by UplinkCoder**
Provides `__ctfe` in D1 as well

### CODE REVIEW

**Comment by mihails-strasuns-sociomantic** [comment]
As usual, please link commit in D2 branch that introduced similar functionality for comparison.

**Comment by stefan-koch-sociomantic** [comment]
@mihails-strasuns-sociomantic Apparently it was introduced in 
https://github.com/dlang/dmd/commit/b0a2add98184c972a0668d460d5d51b7eec37c8c

**Comment by ibuclaw** [comment]
@stefan-koch-sociomantic - That is not the D2 branch?

**Comment by ibuclaw** [comment]
This change looks about right anyway for `__ctfe` support.  Generates 1 during interpret, and 0 during codegen.

File: src/declaration.c [hunk 1]
```diff
@@ -1555,6 +1555,10 @@ void VarDeclaration::checkNestedReference(Scope *sc, Loc loc)
 
 int VarDeclaration::isDataseg()
 {
+    if (ident == Id::ctfe)
+    {
+        return true;
+    }
```
Comment (ibuclaw) [inline src/declaration.c]: Except this, why is this needed?

**Review by ibuclaw — CHANGES_REQUESTED** [review]

File: src/declaration.c (hunk 1 above)
Comment (stefan-koch-sociomantic) [reply src/declaration.c]: because otherwise the error handling below this is triggered.

Comment (ibuclaw) [reply src/declaration.c]: Doesn't that indicate that something is missing though?  This variable should never be written to the data segment.

Comment (stefan-koch-sociomantic) [reply src/declaration.c]: it won't be.
Because it does not exist.
e2ir transforms the variable into a literal `0`.

Comment (mihails-strasuns-sociomantic) [reply src/declaration.c]: Maybe better make it assert in that case?

Comment (stefan-koch-sociomantic) [reply src/declaration.c]: Where would I assert ? `__ctfe` is not a real variable hence it does not have complete type information.
Because it does not have complete type information it would error in `isDataseg` if we would not intercept it.

**Comment by stefan-koch-sociomantic** [comment]
wait ... why does @UplinkCoder want to merge it ?
I should want to merge it ?

**Comment by Burgos** [comment]
Original use case comes from `enforce` not being CTFE-able IIRC. So, piece of code that in some cases _could_ be executed in CTFE, suddenly couldn't, because `enforce` is not allowed to be used. In that case it's possible to `if (__ctfe) assert(...); else enforce(...);`, but sadly not in D2.

Of course, this doesn't answer more philosophical question _why_ do we want this in D1 :-)

**Comment by mihails-strasuns-sociomantic** [comment]
@leandro-lucarella-sociomantic basically combination of very low effort (most of implementation code is already present) and large potential benefit (enables compatible D1/D2 patterns like @nemanja-boric-sociomantic has explained)

**Comment by leandro-lucarella-sociomantic** [comment]
OK, I would be very careful in backporting features unless there is a very good need for them, because it might be hard to asses it doesn't introduce any regressions. At this stage I think we need to focus in moving to D2 rather than backporting features to D1.
